### PR TITLE
Export new generic size_sum()

### LINKED ISSUE #173: `pillar_shaft.vctrs_list_of()` has no customization point
I was planning on creating a `vctrs_list_of` class where each element would be a tibble containing "notes" (i.e. errors and warnings) that happened during model tuning. I would like to be able to control the printing enough to have an end result like:

```r
# A tibble: 2 x 1
            x
  <list<tbl_notes>>
1       [1 W / 2 E]
2       [0 W / 3 E]
```

After looking at this issue https://github.com/r-lib/pillar/issues/137, I was hoping that `pillar_shaft.vctrs_list_of()` would help.

But the current implementation of `pillar_shaft.vctrs_list_of()` provides no customization points for me.

https://github.com/r-lib/pillar/blob/82370d73e9f5b96689454fab675087d0ef0993c9/R/vctrs.R#L18-L23

Is there any hope that we could go back and do what #137 suggests by calling `map(x, obj_sum)`? Then I could control `obj_sum.tbl_notes`

### PR DESCRIPTION
**Pull request by krlmlr**
Closes #173.

### CODE REVIEW

File: tests/testthat/test-format_list_of.R [hunk 1]
```diff
@@ -0,0 +1,4 @@
+test_that("smoke test", {
+  v <- vctrs::new_list_of(list(1L, 2:4), integer())
```
Comment (DavisVaughan) [inline tests/testthat/test-format_list_of.R:2]: I would add another test that also uses `new_list_of()` where one of the elements has a custom `size_sum()` method. That would enforce that you are indeed calling `size_sum()` in `pillar_shaft.vctrs_list_of()`, as opposed to what was being done previously.

File: tests/testthat/test-obj-sum.R [hunk 2]
```diff
@@ -19,7 +19,7 @@ test_that("data frames and common data vectors have size summary", {
     obj_sum <- obj_sum(x)
     ptype <- type_sum(x)
 
-    expect_equal(obj_sum, !! paste0(ptype, size_sum(x)))
+    expect_equal(obj_sum, !! paste_with_space_if_needed(ptype, size_sum(x)))
```
Comment (DavisVaughan) [inline tests/testthat/test-obj-sum.R:22]: Somewhere in this `test-obj-sum.R` file, I would probably also add a test that shows the result of calling `obj_sum()` on an object with a custom `size_sum()` method.

**Review by DavisVaughan — COMMENTED** [review]
I think this looks good! Here is an example of how I could use this to solve #173, using either a list or a list-of. I think my suggested tests to add would catch both of these cases

``` r
library(pillar)
library(tibble)
library(vctrs)

tbl1 <- new_tibble(
  list(
    type = c("warning", "error"), 
    message = c("foo", "bar")
  ), 
  nrow = 2L, 
  class = "tbl_notes"
)

tbl2 <- new_tibble(
  list(
    type = c("warning", "error", "error"), 
    message = c("foo", "bar", "baz")
  ), 
  nrow = 3L, 
  class = "tbl_notes"
)

vec_ptype_abbr.tbl_notes <- function(x, ...) {
  "tbl-notes"
}
type_sum.tbl_notes <- function(x) {
  "tbl-notes"
}
size_sum.tbl_notes <- function(x) {
  w <- sum(x$type == "warning")
  e <- sum(x$type == "error")
  paste0("[", w, " W / ", e, " E]")
}

col1 <- list(tbl1, tbl2)
col2 <- list_of(tbl1, tbl2)

tibble(col1 = col1, col2 = col2)
#> # A tibble: 2 x 2
#>   col1                                 col2
#>   <list>                  <list<tbl-notes>>
#> 1 <tbl-notes [1 W / 1 E]>       [1 W / 1 E]
#> 2 <tbl-notes [1 W / 2 E]>       [1 W / 2 E]
```

**Comment by DavisVaughan** [comment]
Potential issue when there is no size and you put it in a list-of

``` r
library(pillar)
library(tibble)
library(vctrs)

x <- structure(1, class = "foo")
lst <- list_of(x, x)

size_sum.foo <- function(x) {
  ""
}

tibble(lst = lst)
#> # A tibble: 2 x 1
#>           lst
#>   <list<foo>>
#> 1            
#> 2
```

<sup>Created on 2021-01-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by krlmlr** [comment]
Should we use a placeholder in `pillar_shaft.vctrs_list_of()` if the size is empty?

**Comment by krlmlr** [comment]
Maybe `size_sum()` should be responsible only for the inner contents, and pillar still provides the brackets? (Or only adds brackets in the list-of case if it's empty?)

**Pull request merged**